src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala

```diff
@@ -198,16 +200,24 @@ private[impl] class LaunchQueueActor(
         case None => sender() ! None
       }
 
-    case Add(app, count) =>
+    case Add(runSpec, count) =>
       import context.dispatcher
 
       async {
-        val instances = 0.until(count).map { _ => Instance.Scheduled(app, Instance.Id.forRunSpec(app.id)) }
-        val start = await(instanceTracker.schedule(instances))
+        val existingReserved = await(instanceTracker.specInstances(runSpec.id))
+          .filter(_.isReserved)
+          .take(count)
+          .map(_.copy(state = InstanceState(Condition.Scheduled, Timestamp.now(), None, None), runSpecVersion = runSpec.version, unreachableStrategy = runSpec.unreachableStrategy))
+          .map(InstanceUpdateOperation.RelaunchReserved)
+        val instancesToSchedule = existingReserved.length.until(count).map { _ => Instance.Scheduled(runSpec, Instance.Id.forRunSpec(runSpec.id)) }
```

Comment:
this is where the race condition is originated :) it would need some more synchronization here to get around that

Reply:
Out of curiosity, why did you add the logic here? There are very few places that use `LaunchQueue.add`: `TaskReplaceActor`, `StartingBehaviour`, `TaskStartActor` and `MarathonScheduleActor`. I think these should decide not the launch queue. We should sync with Matthias on this.

Overall the launch queue should go and not be extended.

Reply:
because this is the only place that is *JUST ONE* and place through all the schedule calls go through. It can be inside instance tracker but that does not seem like a good place.

I think it is not preventing us to get rid of this guy. We will just need to call this code (or copy this code) to all the places that call LaunchQueue.Add right now...

Reply:
I don't know. This seems like a short term thing.